### ISSUE TITLE
S3 — The BatchNorm recalibration pass is not clean [Epic #353]

### ISSUE THREAD

**Issue by Polichinel** [views-platform/views-hydranet#356]
**Epic:** #353 · **Depends on:** none · **Blocks:** S4

## Background

**C-184** exists because the training floor was seed-bimodal (~40% bad basins) and the root cause was
recurrent BatchNorm. The mitigation is a post-training recalibration pass that recomputes BN running
statistics forward-only, with `momentum=None` (equal weight, no EMA discount). **Those buffers ship
inside every artifact and move the forecast.**

**C-328** is the record of that pass being polluted by training augmentations — four instances to
date. The most recent fix (`training_engine.py:913-929`) introduced `training_augmentation: bool`
and gates `random_flips` and input noise on it, with the comment at `:924`:

> *"`training_augmentation` gates EVERY training-only augmentation, not one of them."*

**It does not.** Two perturbations still reach a recalibration pass.

### (a) Dropout is active on both recalibration paths

`train()` opens with `ctx.model.train()` at `:915`, **unconditionally**, regardless of
`training_augmentation`. The architecture's dropout sites are `LockedDropout` instances
(`HydraBNrecurrentUnet_06_LSTM4.py:361`), which are identity in eval and masking in train
(`locked_dropout.py:13-15`).

So dropout fires during:
- `_recalibrate_bn:1122` — which calls `train(..., training_augmentation=False)` at `:1146` under
  `torch.no_grad()`, and whose comment at `:1142-1145` claims *"Every training-only augmentation is
  suppressed"*
- the `bn_recal_from` lesson loop at `:1325`

**14 of the 15 BatchNorms sit downstream of a dropout site** (`bn_enc_conv1` consumes
`pool0(dropout['e0s'](...))`). So statistics are recomputed with `momentum=None` on activations
inflated by dropout's `1/(1-p)` scaling, while inference runs with dropout off.

**The repository already ships the exact tool.** `_batchnorm_eval:162` is a context manager whose
docstring reads *"`model.eval()` on the whole model would also disable dropout and change the
forward; this flips only the `_BatchNorm` modules that are currently training"* — i.e. it separates
BN state from dropout state. Its **only** use in the repo is the pushforward auxiliary forward at
`:806`. Neither recalibration path uses the inverse.

### (b) Scheduled sampling runs live on the `bn_recal_from` path

`training_engine.py:1325-1338`:

```python
losses = train(
    ctx, sample_handler, pbar, stage_label=slbl,
    ss_epsilon=ss_epsilon,          # <-- passed unconditionally
    fed_grad_sink=_fed_sink,
    training_augmentation=not _bn_recal,
)
```

`ss_epsilon` is computed at `:1295` with no `_bn_recal` term. So on any SS or ITF arm, the recal pass
accumulates BN statistics while `_process_sequence:517` substitutes the model's own predictions for
ground truth at rate ε — and for `ss_reverse=True` (#287) ε is **at `ss_epsilon_max` from lesson 0**.
The two augmentation families are gated inconsistently in the same call.

## ⚠️ needs-decision — this is the only behavioural change in the epic

**Fixing this changes future artifacts.** Any model retrained afterwards carries different BatchNorm
buffers than the same config trained before it. That is exactly what **C-328**'s own fix did, and
what PR #351 initially claimed it had not done until the review corrected it.

The roster was retrained on 2026-09-07/08. Landing this invalidates the comparability of anything
retrained after it against those artifacts.

**Decide and record before implementing:** fix now and accept that the eight roster artifacts become
a pre-fix cohort; or fix behind a flag; or defer to the next retraining cycle. **Do not land it
silently** — write the consequence into the register (C-328) first, the way the release PR's
correction had to be written after the fact.

## Work

- [ ] **Record the decision and the artifact consequence in C-328 BEFORE the code change**
- [ ] (a) Suppress dropout on both recalibration paths — the inverse of `_batchnorm_eval`, keeping BN
      in train mode while putting `LockedDropout` in eval. Do not use `model.eval()`, which would
      also freeze the BN statistics the pass exists to recompute.
- [ ] (b) Gate `ss_epsilon` on `training_augmentation` at `:1325-1338`, or state in the comment why
      scheduled sampling is deliberately exempt
- [ ] Correct `_recalibrate_bn`'s docstring at `:1142-1145` — it currently claims completeness it does
      not have
- [ ] A test that a recalibration forward runs with every `LockedDropout` in eval

## Acceptance criteria

- [ ] Both recalibration paths run with dropout inactive and BN still accumulating
- [ ] `ss_epsilon` is either suppressed on a recal pass or its exemption is justified in the comment
- [ ] The `training_augmentation` comment at `:924` is true of the code, or amended
- [ ] The artifact consequence is in C-328 **before** the change lands, with the dates of the affected
      cohort
- [ ] **The test fails when the fix is reverted** (**C-329**)

## Testing / validation

- `pytest tests/train/test_bn_recal_augmentation.py` plus the new dropout case
- Assert directly on module state during the pass: every `_BatchNorm` in `training` mode, every
  `LockedDropout` not — an assertion on behaviour, not on a call count
- ⚠️ Do **not** assert byte-identity against a pre-fix artifact — the whole point is that it changes
- `ruff check .` and the full suite

## Files

- `views_hydranet/train/training_engine.py` (`:162` `_batchnorm_eval`, `:911-930`, `:1122-1149`,
  `:1295`, `:1325-1338`)
- `views_hydranet/architectures/locked_dropout.py`
- `tests/train/test_bn_recal_augmentation.py`
- `docs/CICs/TrainingEngine.md` — the recalibration invariant
- `reports/technical_risk_register.md` — C-328

**Comment by Polichinel** [comment]
## Implemented — the `needs-decision` is taken: **fix now, and record the artifact consequence first**

This is the epic's **only behavioural change**. The decision and its consequence went into the
register (**C-328 instance 5**) *before* the code changed, per this story's own criteria.

### What was leaking

`training_augmentation`'s comment claimed it gates *"EVERY training-only augmentation"*. Two still
reached both recal paths:

* **Dropout.** `train()` called `ctx.model.train()` unconditionally, so every `LockedDropout` site
  fired. **14 of the 15 BatchNorms sit downstream of one.**
* **Scheduled sampling.** `ss_epsilon` was passed live into the `bn_recal_from` lesson loop.

### The fix

* After `ctx.model.train()`, when `not training_augmentation`: every `nn.Dropout`, `nn.Dropout2d`,
  `nn.Dropout3d` and `LockedDropout` module is put in `eval()`. BatchNorm stays in train mode — it
  has to, that is the point of the pass.
* `ss_epsilon=0.0 if _bn_recal else ss_epsilon` in the `bn_recal_from` loop.

### Tests — `tests/train/test_bn_recal_augmentation.py`

A spy on the dropout modules' training flag: `test_no_recalibration_forward_runs_with_dropout_live`,
`test_ordinary_training_calls_DO_keep_dropout_live` (anti-vacuity),
`test_batchnorm_still_accumulates_while_dropout_is_off`, `test_a_bn_recal_run_passes_zero_epsilon`.

**The last one exists because the deletion test caught me.** The first draft had the `ss_epsilon`
gate with no test; reverting it left everything green — a C-329 instance in my own work. Now
recorded under C-329 with the others.

### The consequence, stated plainly

The eight roster artifacts already trained are files and do not change. **An artifact retrained
after this is not bit-comparable with one trained before it**, because the BN running statistics
written into it are computed on clean inputs. Same class as C-328's earlier instances; the roster
was retrained days ago on the old behaviour.

**Comment by Polichinel** [comment]
## ⚠️ Correction — the dropout half of this story was wrong and is reverted

My comment above says dropout was suppressed on the recal passes and that this changes future
artifacts. **The premise was false.** Production inference is `model.eval()` +
`set_locked_dropout(True)` (ADR-057) — MC-dropout **on**, with a locked mask. So BatchNorm must
normalise dropout-shaped activations, and the C-184 fix that flipped 6/6 bad seeds was validated
with dropout on.

`/code-review max` on #372 measured the draft on the real `HydraBNUNet06_LSTM4` at the fleet's
`p=0.125`: pre-S3 recal `running_var` matched the inference-regime variance to ±0.7% at all 15
layers; the draft's was **14–21% low at the 14 layers downstream of a dropout site**, and the
inference gate-logit std rose 0.396 → 0.450. It moved the buffers *toward* the C-184 bad basin.

**Reverted.** A forward-time hook test now pins dropout ON during every recal forward (the earlier
spies read state after `train()` returned, which a mutation moving the gate past the forward
survived).

**What stands:** the scheduled-sampling half. Now zeroed where ε is *computed* — so the value logged
to wandb is the value used (the draft zeroed only the kwarg and logged the schedule's live ε) — and
`train()` zeroes it again internally whenever `training_augmentation=False`, so the flag means what
it says inside the function, not only at one call site.

**Consequence:** with only the SS half landing, **no roster artifact's buffers change** — none uses
`bn_recal_from` with an active schedule. The "pre-fix cohort" in C-328 instance 5 does not come into
being; the register carries the half-retraction, and the ledger and CIC no longer call S3 the
epic's one behavioural change.

**Comment by Polichinel** [comment]
Merged to `development` in PR #372 (`1d6cef8`). Closed by hand — merges to `development` do not auto-close.
